File: ios_app/GenAIAcademyDashboard/ViewModels/DashboardViewModel.swift
//
//  DashboardViewModel.swift
//  GenAIAcademyDashboard
//
//  ViewModel for dashboard screen
//

import Foundation
import SwiftUI

@MainActor
class DashboardViewModel: ObservableObject {
    @Published var summary: DashboardSummary?
    @Published var charts: DashboardCharts?
    @Published var isLoading = false
    @Published var errorMessage: String?
    @Published var selectedPeriod: Period = .thirtyDays
    
    enum Period: String, CaseIterable {
        case sevenDays = "7d"
        case thirtyDays = "30d"
        case ninetyDays = "90d"
        
        var displayName: String {
            switch self {
            case .sevenDays: return "Last 7 Days"
            case .thirtyDays: return "Last 30 Days"
            case .ninetyDays: return "Last 90 Days"
            }
        }
    }
    
    private let networkManager = NetworkManager.shared
    
    /// Load dashboard summary data
    func loadSummary() async {
        do {
            let endpoint = "\(Constants.dashboardSummaryEndpoint)?period=\(selectedPeriod.rawValue)"
            summary = try await networkManager.request(
                endpoint: endpoint,
                responseType: DashboardSummary.self
            )
            errorMessage = nil
        } catch let error as NetworkError {
            errorMessage = error.errorDescription
        } catch {
            errorMessage = "Failed to load dashboard data"
        }
    }
    
    /// Load dashboard charts data
    func loadCharts() async {
        do {
            let endpoint = "\(Constants.dashboardChartsEndpoint)?period=\(selectedPeriod.rawValue)"
            charts = try await networkManager.request(
                endpoint: endpoint,
                responseType: DashboardCharts.self
            )
        } catch {
            // Charts are optional, don't show error if they fail
            print("Failed to load charts: \(error)")
        }
    }
    
    /// Load all dashboard data
    func loadDashboardData() async {
        isLoading = true
        errorMessage = nil
        
        await withTaskGroup(of: Void.self) { group in
            group.addTask { await self.loadSummary() }
            group.addTask { await self.loadCharts() }
        }
        
        isLoading = false
    }
    
    /// Refresh dashboard data
    func refresh() async {
        await loadDashboardData()
    }
    
    /// Change time period and reload data
    func changePeriod(_ period: Period) async {
        selectedPeriod = period
        await loadDashboardData()
    }
}

File: flutter_app/lib/viewmodels/dashboard_viewmodel.dart
import 'package:flutter/foundation.dart';
import '../models/dashboard_summary.dart';
import '../models/dashboard_charts.dart';
import '../services/network_service.dart';
import '../utils/constants.dart';

/// ViewModel for dashboard screen (MVVM pattern)
class DashboardViewModel extends ChangeNotifier {
  final _networkService = NetworkService();
  
  DashboardSummary? _summary;
  DashboardCharts? _charts;
  bool _isLoading = false;
  String? _errorMessage;
  Period _selectedPeriod = Period.thirtyDays;
  
  DashboardSummary? get summary => _summary;
  DashboardCharts? get charts => _charts;
  bool get isLoading => _isLoading;
  String? get errorMessage => _errorMessage;
  Period get selectedPeriod => _selectedPeriod;
  
  /// Time period enum
  enum Period {
    sevenDays('7d', 'Last 7 Days'),
    thirtyDays('30d', 'Last 30 Days'),
    ninetyDays('90d', 'Last 90 Days');
    
    final String value;
    final String displayName;
    const Period(this.value, this.displayName);
  }
  
  /// Load dashboard summary data
  Future<void> loadSummary() async {
    try {
      final endpoint = '${Constants.dashboardSummaryEndpoint}?period=${_selectedPeriod.value}';
      _summary = await _networkService.get<DashboardSummary>(
        endpoint,
        (json) => DashboardSummary.fromJson(json),
      );
      _errorMessage = null;
      notifyListeners();
    } catch (e) {
      _errorMessage = e.toString().replaceAll('NetworkError: ', '');
      notifyListeners();
    }
  }
  
  /// Load dashboard charts data
  Future<void> loadCharts() async {
    try {
      final endpoint = '${Constants.dashboardChartsEndpoint}?period=${_selectedPeriod.value}';
      _charts = await _networkService.get<DashboardCharts>(
        endpoint,
        (json) => DashboardCharts.fromJson(json),
      );
    } catch (e) {
      // Charts are optional, don't show error if they fail
      print('Failed to load charts: $e');
    }
  }
  
  /// Load all dashboard data
  Future<void> loadDashboardData() async {
    _isLoading = true;
    _errorMessage = null;
    notifyListeners();
    
    await Future.wait([
      loadSummary(),
      loadCharts(),
    ]);
    
    _isLoading = false;
    notifyListeners();
  }
  
  /// Refresh dashboard data
  Future<void> refresh() async {
    await loadDashboardData();
  }
  
  /// Change time period and reload data
  Future<void> changePeriod(Period period) async {
    _selectedPeriod = period;
    await loadDashboardData();
  }
}
